### ISSUE TITLE
teams_workflow 

### ISSUE THREAD

**Issue by vanurag63** [louislam/uptime-kuma#4934]
### ⚠️ Please verify that this question has NOT been raised before.

- [X] I checked and didn't find similar issue

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/uptime-kuma/security/policy)

### 📝 Describe your problem

i unable to connect my webhook with the workflow. the teams is asking for the adptive card. can you give me the steps to do this. 

i have found in internet but unable to get the solution.

### 🐻 Uptime-Kuma Version

Version: 1.23.3 Frontend Version: 1.23.3

### 💻 Operating System and Arch

ubuntu

### 🌐 Browser

chrome

### 🖥️ Deployment Environment

- Runtime: Docker 20.10.9
- Database: sqlite
- Filesystem used to store the database on: ubuntu
- number of monitors: 50+

**Comment by vguttmann** [comment]
The notification for Teams is intended to be used for webhooks created for Teams channels, not for Workflows.
However, Microsoft is retiring Connectors in Teams, which is needed for channel webhooks.

Currently, you need to define a custom data schema in the generic webhook notification

**Comment by vanurag63** [comment]
can you help me with that. can you provide me the documentation. im unable to define the schema

**Comment by vguttmann** [comment]
I don't know the schema myself, but I have opened a feature request to hopefully get something working - I'm also just a user who is annoyed with Microsoft retiring the Teams Webhooks
https://github.com/louislam/uptime-kuma/issues/4937

**Comment by vanurag63** [comment]
@vguttmann is there any other option for now to migrate to workflow.

**Comment by vguttmann** [comment]
I don't know. Existing Webhooks will continue working until October, and you can still create new webhooks.
In the meantime, there should be a PR that fixes this, I think

**Comment by vanurag63** [comment]
@vguttmann after october there will be a fix. let hope.

**Comment by vanurag63** [comment]
@louislam can you give update on this

**Comment by CommanderStorm** [comment]
See my comment from https://github.com/louislam/uptime-kuma/pull/4538#issuecomment-2227995155

> Using Workflows instead of webhooks also means changing the bits of documentation surrounding that.
If you want to rebase this change, the [contribution guide is here](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md)
> 
> ![image](https://github.com/user-attachments/assets/e2f043d5-b8d2-434b-9525-3b837bbde142)

=> instead of expecting other people to do the work, consider helping out

**Comment by Tabisch** [comment]
I made a PR to fix this in the project directly.
For the interim I've made a docker image that converts the calls to the new workflow webhooks so we have something that works until there is a proper fix.
https://github.com/Tabisch/uptime-kuma-teams-webhook-shim

**Comment by tracyhoward** [comment]
> See my comment from [#4538 (comment)](https://github.com/louislam/uptime-kuma/pull/4538#issuecomment-2227995155)
> 
> > Using Workflows instead of webhooks also means changing the bits of documentation surrounding that.
> > If you want to rebase this change, the [contribution guide is here](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md)
> > ![image](https://private-user-images.githubusercontent.com/26258709/348646701-e2f043d5-b8d2-434b-9525-3b837bbde142.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIyNjc3MjEsIm5iZiI6MTcyMjI2NzQyMSwicGF0aCI6Ii8yNjI1ODcwOS8zNDg2NDY3MDEtZTJmMDQzZDUtYjhkMi00MzRiLTk1MjUtM2I4MzdiYmRlMTQyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI5VDE1MzcwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExYWQ0N2EwNjg3YWQxODVjNzQ3NWQ3ZTcxYTk1ZTU5MGJiZTI1NjMwNzcyZTFiNjBjZTMyYmUxNmZkYTYxODYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5njbEFabQubzdxohV7idrvxWQtM37Rx1a06HpQ1XZK8)
> 
> => instead of expecting other people to do the work, consider helping out

We are really only in need of documentation updates so that the changes can be merged into a "patch" release; correct?  If that is true; i will seek out some bandwidth to document/screenshot new PowerAutomate Workflow basic guidance

**Comment by CommanderStorm** [comment]
Help documenting this would be appreciated ^^

**Comment by kaaac** [comment]
Workflows webhook requires receiving a payload with attachments in the format:

```
{
    "type": "message",
    "attachments": []
}
```
And each of the attachments must be compliant with the Adaptive Cards schema. For example:

```
{
    "type": "message",
    "attachments": [
        {
            "contentType": "application/vnd.microsoft.card.adaptive",
            "content": {
                "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
                "type": "AdaptiveCard",
                "version": "1.3",
                "body": [
                    {
                        "type" : "Container",
                        "showBorder" : true,
                        "roundedCorners" : true,
                        "items" : [
                            {
                                "type" : "TextBlock",
                                "text" : "Example Message",
                                "weight" : "Bolder",
                                "wrap" : "true"
                            }
                        ]
                    }
                ]
            }
        }
    ]
}

```
The webhook notification only allows me to configure payload when selecting "Request Body" option "Custom Body".

![image](https://github.com/user-attachments/assets/f023207d-e302-47d8-80dd-e827941f901d)


Unfortunately, when I try to send I get the following message:

```
Error: AxiosError: Request failed with status code 400 {"error":{"code":"InvalidRequestContent","message":"The input body for trigger 'manual' of type 'Request' must be of type JSON, but was of type 'application/x-www-form-urlencoded'."}}
```

When selecting the remaining options for "Request body" there is no possibility to define your own schema.

![image](https://github.com/user-attachments/assets/45483499-cd54-401a-93b1-ca2950146ab4)

Maybe someone has an idea how to get around this?

Cheers

**Comment by meako689** [comment]
@kaaac 

![image](https://github.com/user-attachments/assets/5495b973-ec1c-4582-baf4-9189b42636e4)

Sending content type as header should do the trick.

**Comment by tracyhoward** [comment]
> Help documenting this would be appreciated ^^

I have created some markdown documentation with screenshots for creating the proper workflow.  I am not a front-end developer and I don't want to waste anybody's time with a PR showing the documentation of how to set this stuff up.  

I really just wanted to provide the basic flow which makes this bugfix work.  Please advise on how you would like me to proceed?  I can paste everything here in a .md post or go through the PR process (with your guidance on the structure of the source files for documentation).

either way will be the same result
thanks

**Comment by tracyhoward** [comment]
## Setup Notifications with Teams, Power Automate and Uptime-Kuma

- Navigate to Power Automate within the Microsoft 365 Admin Center https://www.microsoft365.com

![microsoft365](https://github.com/user-attachments/assets/259bdf99-733a-4034-9fbc-1dd630788d64)

- Select "Templates" from the left-hand menu
- Type "When a Teams Webhook request is received" into the template search bar
- Select "Post to a channel when a webhook request is received"
    
![templateCreation](https://github.com/user-attachments/assets/e8bba543-5ee5-4f78-a17d-37e74f6a3c15)

- Setup a Teams Connection (with your credentials or a service account credentials) and "Continue"

![connectionContinue](https://github.com/user-attachments/assets/ffe797b7-7bf0-4565-a84e-9b3735cde868)

- Wait for the connection to be established (5-10 seconds)
- Select Team and Channel to post to then select "Create"
    
![editContinue](https://github.com/user-attachments/assets/2ab19879-deb1-4c82-bdac-caaa0f0ae189)

- Open/Edit your new flow so that you can retrieve the webhook URL to used from Uptime-Kuma

![editNewFlow](https://github.com/user-attachments/assets/83939c9b-6b23-48de-ba8c-6c54fa0a6597)

![webhookUrl](https://github.com/user-attachments/assets/e1e7301c-cb2a-4432-b9d1-5acf674c4e94)

<i> if using your personal account you may need to make an additional edit as seen in the following steps </i>

- select the "Post card in a chat or channel portion of the flow and edit the "Post As" value from "Flow Bot" to "User"
    
![finalEdit](https://github.com/user-attachments/assets/47ec512f-66fd-4e77-b13d-cd7dd3fb4ba0)

That should allow for you to set the webhook URL in Uptime-Kuma to post to the Teams channel of your choice.

here's a curl command that you can use to test:
```
curl -X POST \
          -H "Content-type: application/json" \
          -d "{
    \"type\": \"message\",
    \"attachments\": [
        {
            \"contentType\": \"application/vnd.microsoft.card.adaptive\",
            \"content\": {
                \"type\": \"AdaptiveCard\",
                \"body\": [
                    {
                        \"type\": \"TextBlock\",
                        \"weight\": \"Bolder\",
                        \"size\": \"Large\",
                        \"color\": \"$txtColor\",
                        \"text\": \"Test Results for $appName\"
                    },
                    {
                        \"type\": \"FactSet\",
                        \"facts\": [
                            {
                                \"title\": \"Test Passed:\",
                                \"value\": \"$passedTests\"
                            },
                            {
                                \"title\": \"Test Failed:\",
                                \"value\": \"$failedTests\"
                            },
                            {
                                \"title\": \"Author:\",
                                \"value\": \"$GITHUB_ACTOR\"
                            },
                            {
                                \"title\": \"Commit:\",
                                \"value\": \"$GITHUB_SHA\"
                            }
                        ]
                    }
                ],
                \"actions\": [                
                    {
                        \"type\": \"Action.OpenUrl\",
                        \"title\": \"View on GitHub\",
                        \"url\": \"google.com\",
                        \"style\": \"positive\"
                    },
                    {
                        \"type\": \"Action.OpenUrl\",
                        \"title\": \"View Report\",
                        \"url\": \"google.com\",
                        \"style\": \"positive\"
                    }
                ]
            }
        }
    ]
}" \
[ YOUR NEW WEBHOOK URL HERE ]
```

**Comment by mofc-foodbank** [comment]
> ## Setup Notifications with Teams, Power Automate and Uptime-Kuma
> * Navigate to Power Automate within the Microsoft 365 Admin Center https://www.microsoft365.com
> 

I followed your guide here, but when I push a test notification from Kuma, I get an error in Power Automate.  On the Flows page, I can see **28 day run history** and it shows a status of **Failed** after I hit the test.  Going into the attempt on Power Automate, it shows the chart-looking thing like in your screenshot (When a Teams... --> Send each adaptive card --> Post card...) and there is a red exclamation mark on the middle "Send each adaptive card."  Clicking it shows the error below: 

> ExpressionEvaluationFailed
> The execution of template action 'Send_each_adaptive_card' failed: the result of the evaluation of 'foreach' expression '@triggerOutputs()?['body']?['attachments']' is of type 'Null'. The result must be a valid array.

Is there a step that I missed or maybe something you left out of your instructions thinking it'd be obvious?

**Comment by tracyhoward** [comment]
> @tracyhoward
> 
> > ## Setup Notifications with Teams, Power Automate and Uptime-Kuma
> > 
> > * Navigate to Power Automate within the Microsoft 365 Admin Center https://www.microsoft365.com
> 
> I followed your guide here, but when I push a test notification from Kuma, I get an error in Power Automate. On the Flows page, I can see **28 day run history** and it shows a status of **Failed** after I hit the test. Going into the attempt on Power Automate, it shows the chart-looking thing like in your screenshot (When a Teams... --> Send each adaptive card --> Post card...) and there is a red exclamation mark on the middle "Send each adaptive card." Clicking it shows the error below:
> 
> > ExpressionEvaluationFailed
> > The execution of template action 'Send_each_adaptive_card' failed: the result of the evaluation of 'foreach' expression '@triggerOutputs()?['body']?['attachments']' is of type 'Null'. The result must be a valid array.
> 
> Is there a step that I missed or maybe something you left out of your instructions thinking it'd be obvious?

ya; this won't work from Kuma at this point...A PR was raised/completed back in march of this year to account for the Teams change; however, that has not made its way into a release at this time.  the instructions above are what you will follow once those changes are in place (a new tagged Kuma release).  The curl command should work for the 'flow' that you created using this documentation.

https://github.com/louislam/uptime-kuma/pull/4538/files

**Comment by CommanderStorm** [comment]
@Tabisch is currently backporting that change here https://github.com/louislam/uptime-kuma/pull/4957

#4538 is a PR raised against master and will not be included in 1.x releases.

**Comment by blakefallow** [comment]
For anyone that is still facing this issue and is unable to use the patched beta version or the SMTP workaround. We were able to use the preinstalled [Apprise](https://github.com/caronc/apprise) integration on v1.23.16.

<img width="908" height="215" alt="Image" src="https://github.com/user-attachments/assets/e5245850-6d22-4773-89c5-c9fd5bf02579" />

**Comment by jamesc-skyward** [comment]
> For anyone that is still facing this issue and is unable to use the patched beta version or the SMTP workaround. We were able to use the preinstalled [Apprise](https://github.com/caronc/apprise) integration on v1.23.16.
> <img alt="Image" width="908" height="215" src="https://private-user-images.githubusercontent.com/10758471/490173235-e5245850-6d22-4773-89c5-c9fd5bf02579.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjEyMzU3MzQsIm5iZiI6MTc2MTIzNTQzNCwicGF0aCI6Ii8xMDc1ODQ3MS80OTAxNzMyMzUtZTUyNDU4NTAtNmQyMi00NzczLTg5YzUtYzlmZDViZjAyNTc5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDIzVDE2MDM1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhOTQ0ZmNhNzA0YTlkNWQxZGMxZjE2NjE4Y2ZkYjI3ZmYyMWRjNTk5YmViM2U2YzRlNzUxMWE3ZWU4ODhhNjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qaL4e0WNZtD0d93wqFqL1J3ApS9HA10rRF4AFA-rtVs">

It looks like this solution no longer works since Microsoft changed their URL pattern. It looks like Apprise addressed this in [version 1.95](https://github.com/caronc/apprise/releases/tag/v1.9.5), but Uptime-Kuma will need to use the updated version.

**Comment by blakefallow** [comment]
> > For anyone that is still facing this issue and is unable to use the patched beta version or the SMTP workaround. We were able to use the preinstalled [Apprise](https://github.com/caronc/apprise) integration on v1.23.16.
> > <img alt="Image" width="908" height="215" src="https://private-user-images.githubusercontent.com/10758471/490173235-e5245850-6d22-4773-89c5-c9fd5bf02579.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjEyMzU3MzQsIm5iZiI6MTc2MTIzNTQzNCwicGF0aCI6Ii8xMDc1ODQ3MS80OTAxNzMyMzUtZTUyNDU4NTAtNmQyMi00NzczLTg5YzUtYzlmZDViZjAyNTc5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDIzVDE2MDM1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhOTQ0ZmNhNzA0YTlkNWQxZGMxZjE2NjE4Y2ZkYjI3ZmYyMWRjNTk5YmViM2U2YzRlNzUxMWE3ZWU4ODhhNjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qaL4e0WNZtD0d93wqFqL1J3ApS9HA10rRF4AFA-rtVs">
> 
> It looks like this solution no longer works since Microsoft changed their URL pattern. It looks like Apprise addressed this in [version 1.95](https://github.com/caronc/apprise/releases/tag/v1.9.5), but Uptime-Kuma will need to use the updated version.

We also eventually hit the same issue. Never had these issues with Slack...

We ended up using the generic webhook with a custom body:

Example:
`{
  "type": "message",
  "attachments": [
    {
      "summary": "{{monitorJSON['name']}} - {{msg}}",
      "contentType": "application/vnd.microsoft.card.adaptive",
      "contentUrl": null,
      "content": {
        "type": "AdaptiveCard",
        "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
        "version": "1.4",
        "body": [
          {
            "type": "ColumnSet",
            "columns": [
              {
                "type": "Column",
                "width": "auto",
                "items": [
                  {
                    "type": "Image",
                    "url": "https://uptimekuma.org/wp-content/uploads/2025/01/Uptime-Kuma-Logo.png",
                    "size": "Small",
                    "altText": "Uptime Kuma"
                  }
                ],
                "verticalContentAlignment": "Center"
              },
              {
                "type": "Column",
                "width": "stretch",
                "items": [
                  { "type": "TextBlock", "text": "Kuma Alert", "size": "Large", "weight": "Bolder" },
                  { "type": "TextBlock", "text": "{{msg}}", "wrap": true, "spacing": "Small" }
                ]
              }
            ]
          },
          {
            "type": "Container",
            "spacing": "Medium",
            "items": [
              { "type": "TextBlock", "text": "{{monitorJSON.name}}", "weight": "Bolder", "size": "Medium" },
              {
                "type": "FactSet",
                "facts": [
                  { "title": "Type",      "value": "{{monitorJSON.type}}" },
                  { "title": "URL",       "value": "{{monitorJSON.url}}" },
                  { "title": "Status",    "value": "{{heartbeatJSON.status}}" },
                  { "title": "Time",      "value": "{{heartbeatJSON.time}}" },
                  { "title": "Ping (ms)", "value": "{{heartbeatJSON.ping}}" },
                  { "title": "Duration",  "value": "{{heartbeatJSON.duration}}" }
                ]
              },
              { "type": "TextBlock", "text": "{{heartbeatJSON.msg}}", "wrap": true, "spacing": "Small", "isSubtle": true }
            ]
          },
          {
            "type": "ActionSet",
            "actions": [
              { "type": "Action.OpenUrl", "title": "Open Endpoint",        "url": "{{monitorJSON.url}}" },
              { "type": "Action.OpenUrl", "title": "Open in Uptime Kuma",  "url": "{INSERT-BASE-URL-HERE}/{{monitorJSON.id}}" }
            ]
          }
        ]
      }
    }
  ]
}`

With the additional header:
`{
"Content-Type": "application/json"
}`
